### PR TITLE
added urlrewrite plugin for rmz.cr (rapidmoviez.com|rapidmoviez.eu)

### PR DESCRIPTION
**Pull request by thawn**
### Motivation for changes:
rmz.cr needs an urlrewrite plugin for the rss feed (rmz.cr/feed) because the feed only provides links back to the rmz.cr release page.

### Detailed changes:
    rmz.cr (rapidmoviez.com) urlrewriter
    Version 0.1
    
    Configuration

    rmz:
      url_re:
        - domain\.com
        - domain2\.org

    Only add links that match any of the regular expressions listed under url_re.

    If more than one valid link is found, the url of the entry is rewritten to
    the first link found. The complete list of valid links is placed in the
    'urls' field of the entry.
    
    Therefore, it is recommended, that you configure your output to use the
    'urls' field instead of the 'url' field.
    
    For example, to use jdownloader 2 as output, you would use the exec plugin:
    exec:
      - echo "text={{urls}}" >> "/path/to/jd2/folderwatch/{{title}}.crawljob"

### Config usage if relevant (new plugin or updated schema):
```
    rmz:
      url_re:
        - domain\.com
        - domain2\.org
```
### Log and/or tests output (preferably both):
```
DEBUG  rmz  TASK  Using links_re filters: <list of filter expressions>
DEBUG  rmz  TASK  Url: "<link found at entry url>" matched filter: <expression>

INFO      rmz  TASK  Found 2 links at <entry url>
```
#### To Do:

- maybe extend this to work also as a search plugin for rmz.cr

### CODE REVIEW

File: flexget/plugins/sites/rmz.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    
+    For example, to use jdownloader 2 as output, you would use the exec plugin:
+    exec:
+      - echo "text={{urls}}" >> "/path/to/jd2/folderwatch/{{title}}.crawljob"
+    """
+
+    schema = {
+        'type': 'object',
+        'properties': {
+            'links_re': {'type': 'array', 'items': {'format': 'regexp'} }
```
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: 1. there's a mismatch between this value and the one in the doc (`url_re`).
2. I'm not sure why is this needed. Don't you know the TLD and format you need for the rewrite? why does the user need to input this? URL rewrite plugin do not need to rely on config.

File: flexget/plugins/sites/rmz.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        for i in reversed(range(len(urls))):
+            urls[i]=urls[i].encode('ascii','ignore')
+            if regexps:
+                accept = False
+                for regexp in regexps:
+                    if re.search(regexp, urls[i]):
+                        accept = True
+                        log.debug('Url: "%s" matched filter: %s', urls[i], regexp)
+                        break
+                if not accept:
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Since you `break` when finding a match, you could just use `for-else` instead of the `accept` variable.

File: flexget/plugins/sites/rmz.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        urls=[]
+        for element in link_elements:
+            urls.extend(element.text.splitlines())
+        regexps = self.config.get('links_re', None)
+        if regexps:
+            log.debug('Using links_re filters: %s', str(regexps))
+        else:
+            log.debug('No link filters configured, using all found links.')
+        for i in reversed(range(len(urls))):
+            urls[i]=urls[i].encode('ascii','ignore')
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Lacking some whitespace around equals and after comma.

File: flexget/plugins/sites/rmz.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                accept = False
+                for regexp in regexps:
+                    if re.search(regexp, urls[i]):
+                        accept = True
+                        log.debug('Url: "%s" matched filter: %s', urls[i], regexp)
+                        break
+                if not accept:
+                    log.debug('Url: "%s" does not match any of the given filters: %s', urls[i], str(regexps))
+                    del(urls[i])
+        numlinks=len(urls)
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Space

File: flexget/plugins/sites/rmz.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        log.info('Found %d links at %s.',numlinks, entry['url'])
+        if numlinks>0:
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Space

File: flexget/plugins/sites/rmz.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            entry.setdefault('urls', urls)
+            entry['url']=urls[0]
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: space

File: flexget/plugins/sites/rmz.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        else:
+            raise UrlRewritingError('No useable links found at %s' % entry['url'])
+
+@event('plugin.register')
+def register_plugin():
+    plugin.register(UrlRewriteRmz, 'rmz', interfaces=['urlrewriter', 'task'], api_ver=2)
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: There should be a newline at the end of the file.

File: flexget/plugins/sites/rmz.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    #grab config
+    def on_task_start(self, task, config):
+        self.config = config
+
+    # urlrewriter API
+    def url_rewritable(self, task, entry):
+        url = entry['url']
+        rewritable_regex = '^https?:\/\/(www.)?(rmz\.cr|rapidmoviez\.(com|eu))\/.*'
+        if re.match(rewritable_regex, url):
```
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: suggestion:
```python
return re.match(rewritable_regex, url) is not None
```

File: flexget/plugins/sites/rmz.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            raise UrlRewritingError(e)
+        try:
+            soup = get_soup(page.text)
+        except Exception as e:
+            raise UrlRewritingError(e)
+        link_elements = soup.find_all('pre', class_='links')
+        urls=[]
+        for element in link_elements:
+            urls.extend(element.text.splitlines())
+        regexps = self.config.get('links_re', None)
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: `None` is already the default value returned by `dict.get()`.

File: flexget/plugins/sites/rmz.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        if regexps:
+            log.debug('Using links_re filters: %s', str(regexps))
+        else:
+            log.debug('No link filters configured, using all found links.')
+        for i in reversed(range(len(urls))):
```
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: i'd use `enumerate` here

File: flexget/plugins/sites/rmz.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            urls[i]=urls[i].encode('ascii','ignore')
+            if regexps:
+                accept = False
+                for regexp in regexps:
```
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: this will crash if `regexps` is `None`

File: flexget/plugins/sites/rmz.py (hunk 4 above)
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: use snake case `num_links`

File: flexget/plugins/sites/rmz.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            return True
+        else:
+            return False
+
+    @plugin.internet(log)
+   # urlrewriter API
+    def url_rewrite(self, task, entry):
+        try:
+            page = task.requests.get(entry['url'])
+        except Exception as e:
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Too broad

File: flexget/plugins/sites/rmz.py (hunk 5 above)
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: not pythonic, no need to check if higher than 0, just do:
```python
if numlinks:
```

File: flexget/plugins/sites/rmz.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            raise UrlRewritingError(e)
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Probably need to wrap `e` in `str()`

File: flexget/plugins/sites/rmz.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        try:
+            soup = get_soup(page.text)
+        except Exception as e:
+            raise UrlRewritingError(e)
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Probably need to wrap `e` in `str()`

File: flexget/plugins/sites/rmz.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    @plugin.internet(log)
+   # urlrewriter API
+    def url_rewrite(self, task, entry):
+        try:
+            page = task.requests.get(entry['url'])
+        except Exception as e:
+            raise UrlRewritingError(e)
+        try:
+            soup = get_soup(page.text)
+        except Exception as e:
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: This one is also too broad, but since BS4 can raise different exceptions it's fine in this case...

File: flexget/plugins/sites/rmz.py (hunk 2 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: Thanks for the tip, I'll try that.

File: flexget/plugins/sites/rmz.py (hunk 11 above)
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: Look at line 90

File: flexget/plugins/sites/rmz.py (hunk 1 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: 1. I'll fix that asap.
2. The link_re config parameter works just like the link_re parameter in the html input plugin (thats why I renamed it).
The reason for using this parameter is that on the rmz.cr site, they often offer links to several filehosters. This parameter allows the user to choose which filehosters he prefers. For example, the user may have an account for one of the filehosters or he may prefer some filehosters because they offer superior download speed or no file size restrictions for free users.

File: flexget/plugins/sites/rmz.py (hunk 12 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: Could you please suggest an error class to use? I looked at the html plugin but they do not catch errors at all.
I tried to use requests.exceptions.RequestException but got an error.

File: flexget/plugins/sites/rmz.py (hunk 1 above)
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: ah, gotcha. rmz.cr is an indexing site and the links are to external sites.
maybe name it `filehoster` instead then?

File: flexget/plugins/sites/rmz.py (hunk 12 above)
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: that is the correct exception. What error did you get? did you import it correctly?

File: flexget/plugins/sites/rmz.py (hunk 11 above)
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: ah ok, maybe just do `regexps = self.config.get('links_re', [])` and avoid the if statement

File: flexget/plugins/sites/rmz.py [hunk 16]
```diff
@@ -60,48 +62,42 @@ def on_task_start(self, task, config):
...
+        for i in reversed(enumerate(urls)):
+            urls[i] = urls[i].encode('ascii', 'ignore')
+            for regexp in regexps:
+                if re.search(regexp, urls[i]):
+                    log.debug('Url: "%s" matched filehoster filter: %s', urls[i], regexp)
+                    break
+            else:
+                log.debug('Url: "%s" does not match any of the given filehoster filters: %s', urls[i], str(regexps))
+                del(urls[i])
+        num_links=len(urls)
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: space

File: flexget/plugins/sites/rmz.py [hunk 17]
```diff
@@ -60,48 +62,42 @@ def on_task_start(self, task, config):
...
+        log.info('Found %d links at %s.',num_links, entry['url'])
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: space after comma

File: flexget/plugins/sites/rmz.py (hunk 10 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I need to go through the array in reversed order because I am removing some elements within the for loop.

I just tried and enumerate did not work well with reversed:
when I tried to use `reversed(enumerate(urls))`, I got the error:
TypeError: argument to reversed() must be a sequence
Then I tried `reversed(list(enumerate(urls)))` (which is imho at least as ugly as what I used originally) but that gave me the error:
TypeError: list indices must be integers, not tuple

so I gave up and reverted to `reversed(range(len(urls)))`

File: flexget/plugins/sites/rmz.py [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+            log.debug('No filehoster filters configured, using all found links.')
+        for i in reversed(range(len(urls))):
+            urls[i] = urls[i].encode('ascii', 'ignore')
+            for regexp in regexps:
+                if re.search(regexp, urls[i]):
+                    log.debug('Url: "%s" matched filehoster filter: %s', urls[i], regexp)
+                    break
+            else:
+                log.debug('Url: "%s" does not match any of the given filehoster filters: %s', urls[i], str(regexps))
+                del(urls[i])
```
Comment (cvium) [inline flexget/plugins/sites/rmz.py]: It's a bad idea to iterate over a list and delete elements from it. You should either create a copy of the list or add to another list instead of deleting.

File: flexget/plugins/sites/rmz.py (hunk 7 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I added a newline in my latest commit, but it does not show up here. Am I missing something?

File: flexget/plugins/sites/rmz.py (hunk 18 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I never experienced any problem when I was traveling the list in reversed order. Are there any performance implications (although we are talking about 100 entries in the list tops)?

File: flexget/plugins/sites/rmz.py [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+        num_links = len(urls)
+        log.info('Found %d links at %s.',num_links, entry['url'])
```
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: I think `log.verbose` level would fit here better

File: flexget/plugins/sites/rmz.py (hunk 1 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: good idea, done in the latest commit.

File: flexget/plugins/sites/rmz.py (hunk 11 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: o.k. got it to work like this.

File: flexget/plugins/sites/rmz.py (hunk 7 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I added a newline a few commits ago and I just checked on my local machine and with the online editor and there is definitely a newline at the end of the file now.

Don't know why this review did not get marked as outdated...

File: flexget/plugins/sites/rmz.py (hunk 12 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I got this fixed now. I had to add `from requests.exceptions import RequestException`. Now I am using RequestException instead of Exception.

File: flexget/plugins/sites/rmz.py (hunk 13 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: done

File: flexget/plugins/sites/rmz.py (hunk 9 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I replaced this with [] to make sure it does not brake my for loop in line 86

File: flexget/plugins/sites/rmz.py (hunk 18 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: I read up on the performance implications of del.
switched to adding to a new list.

File: flexget/plugins/sites/rmz.py (hunk 19 above)
Comment (thawn) [inline flexget/plugins/sites/rmz.py]: changed that and added the space after the comma.

**Comment by thawn** [comment]
@cvium @liiight 
Thank you so much for the kind code review. I learned a lot today!
I think I now managed to address all the issues you raised. Let me know if there is more that needs to be improved.

Cheers,

Thawn

File: flexget/plugins/sites/rmz.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+            for regexp in regexps:
+                if re.search(regexp, urls[i]):
+                    filtered_urls.append(urls[i])
+                    log.debug('Url: "%s" matched filehoster filter: %s', urls[i], regexp)
+                    break
+            else:
+                log.debug('Url: "%s" does not match any of the given filehoster filters: %s', urls[i], str(regexps))
+        if regexps:
+            log.debug('Using filehosters_re filters: %s', str(regexps))
+            urls=filtered_urls
```
Comment (liiight) [inline flexget/plugins/sites/rmz.py]: Missing spaces. See about running pep8 on your code. You can also use pycharm which is free

Comment (thawn) [inline flexget/plugins/sites/rmz.py]: Thanks for the tip. I used autopep8 (via eclipse).

**Comment by liiight** [comment]
Thanks, care to update the wiki?

**Comment by thawn** [comment]
@liiight Sure, I'll write a wiki page and link to it under the "Data operations" section next to the generic urlrewrite plugin.

Thanks again for your help with improving the code!

**Comment by thawn** [comment]
Wiki page is done:
https://www.flexget.com/Plugins/rmz

and linked from the plugins page (section Modification/Data Operations):
https://flexget.com/Plugins#modification
